File: scanner/internal/collect/discovery.go
// SPDX-License-Identifier: MIT
// Copyright (C) 2026 Bijstaan

package collect

import (
	"github.com/bijstaan/glpi-netscan/internal/payload"
	"github.com/bijstaan/glpi-netscan/internal/snmpx"
)

// Discovery: the cheap pass that decides whether the expensive one is needed.
//
// Measured against the SNMP simulators, a full inventory walk costs
// between 86 and 157 times what a probe does — roughly 2ms against 200-360ms,
// and that is with tiny tables on loopback. A 48-port switch across a WAN is
// far worse. So the cost of a sweep is dominated by full-walking the devices
// that *answer*, not by probing the addresses that do not, and the saving comes
// from not re-walking a device that has not changed since the last time.
//
// Which means discovery has to carry enough for someone to make that decision.
// These three counters are what the standard MIBs offer:
//
//   - ifTableLastChange   an interface was created or deleted
//   - ifStackLastChange   the layering between interfaces changed
//   - entLastChangeTime   the physical inventory changed — a module, a stack
//     member, a transceiver
//
// All three are sysUpTime-relative, so they are meaningless across a reboot —
// which does not matter, because a reboot is itself a reason to re-walk and is
// visible as uptime going backwards.
//
// What they deliberately do not catch is a port changing state, or a MAC moving
// between ports. Those change constantly and would defeat the whole point. The
// bound on that staleness is the inventory interval, not these counters.
const (
	oidIfTableLastChange = "1.3.6.1.2.1.31.1.5.0"
	oidIfStackLastChange = "1.3.6.1.2.1.31.1.6.0"
	oidEntLastChangeTime = "1.3.6.1.2.1.47.1.4.1.0"
)

// Markers reads the change counters, in one GET alongside the probe.
//
// Absent counters come back as zero, which is correct: a device that does not
// implement them offers no evidence of change, so the decision falls back to
// the interval. Errors are swallowed for the same reason — a missing scalar is
// the normal case on simple hardware, not a fault worth failing a sweep over.
func Markers(sess *snmpx.Session, ident Identity) payload.Markers {
	markers := payload.Markers{UptimeTicks: ident.UpTimeTicks}

	res, err := sess.Get([]string{
		oidIfTableLastChange,
		oidIfStackLastChange,
		oidEntLastChangeTime,
	})
	if err != nil {
		return markers
	}

	markers.IfTableLastChange = num(res, oidIfTableLastChange)
	markers.IfStackLastChange = num(res, oidIfStackLastChange)
	markers.EntityLastChange = num(res, oidEntLastChangeTime)

	return markers
}

// DiscoveryPayload builds the submission for the cheap pass.
//
// Deliberately thin. It carries what a discovery is *for* — something answers
// here, this is what it says it is — and the markers that let the server decide
// whether the expensive pass is warranted. No port walk, no tables.
//
// The device type is included because GLPI files a discovery into an asset type
// too, and classifying from sysServices and a couple of cheap probes is what
// stops every discovered device landing as Unmanaged and then being corrected
// on the next full walk.
func DiscoveryPayload(
	deviceID, host string,
	ident Identity,
	dev *payload.Device,
	markers payload.Markers,
	versionClient string,
) payload.Inventory {
	return payload.Inventory{
		DeviceID:    deviceID,
		Action:      string(payload.QueryDiscovery),
		ItemType:    payload.GLPIItemType(dev.Type),
		ScanAddress: host,
		Markers:     &markers,
		Content: payload.Content{
			Device:        dev,
			VersionClient: versionClient,
		},
	}
}

// DiscoveryDevice builds the identity block from a probe alone.
//
// No profile OIDs are read: those are GETs against a device we may be about to
// decide not to walk, and the whole point of this pass is that it costs almost
// nothing. Serial and model are left to the inventory pass.
func DiscoveryDevice(ident Identity) *payload.Device {
	dev := &payload.Device{
		Type:        classifyFromIdentity(ident),
		Name:        ident.SysName,
		Description: ident.SysDescr,
		Contact:     ident.SysContact,
		Location:    ident.SysLocation,
	}

	if dev.Name == "" {
		dev.Name = ident.SysName
	}

	return dev
}

// classifyFromIdentity types a device without walking anything.
//
// Only sysServices is available this cheaply, so this is coarser than the full
// classification — a printer looks like any other IP appliance here. That is
// the right trade: a discovery says "something is at this address and it speaks
// SNMP", and the inventory pass corrects the type when it runs. Guessing harder
// from sysObjectID would mean shipping a vendor table that goes stale.
func classifyFromIdentity(ident Identity) string {
	// datalink(2) plus internet(4) is a switch or router; the bits are a
	// bitmask, not an enumeration.
	const (
		datalink = 0x02
		internet = 0x04
	)

	if ident.SysServices&datalink != 0 && ident.SysServices&internet != 0 {
		return payload.TypeNetworking
	}

	return payload.TypeUnmanaged
}

File: src/Discovery.php
<?php
/*
 * SPDX-License-Identifier: GPL-3.0-or-later
 * Copyright (C) 2026 Bijstaan
 */

namespace GlpiPlugin\Glpinetscan;

use DBmysql;

/**
 * The discovery/inventory split: deciding which devices are worth walking.
 *
 * A full inventory walk costs between 86 and 157 times what a discovery probe
 * does — measured against the SNMP simulators, with tiny tables over
 * loopback; a 48-port switch across a WAN is far worse. So a sweep's cost is
 * dominated by fully walking the devices that answer, not by probing the
 * addresses that do not, and the saving comes from not re-walking a device that
 * has not changed.
 *
 * The decision lives here, on the server, rather than on the scanner. GLPI is
 * the only party that knows what it already holds, when it last held it, and
 * whether someone has since deleted the asset — a scanner keeping its own notes
 * would happily skip a device whose asset had been removed, and GLPI would stay
 * empty with nothing to explain why.
 *
 * A device is walked when any of these is true:
 *
 *   - it has never been walked, or the asset it produced no longer exists
 *   - it rebooted since the last walk (uptime went backwards)
 *   - a standard change counter moved: an interface was created or deleted,
 *     the layering between interfaces changed, or the physical inventory did
 *   - the last walk is older than the target's inventory interval
 *
 * That last rule is what bounds the staleness the counters cannot see. None of
 * them moves when a port goes down or a MAC moves between ports, and both of
 * those matter to GLPI — so "nothing looks changed" is never allowed to mean
 * "never walk again".
 */
final class Discovery
{
    public const TABLE = 'glpi_plugin_glpinetscan_seen';

    /**
     * Record a discovery submission and decide whether a full walk is wanted.
     *
     * @param array<string,mixed> $payload one discovery submission
     * @return bool whether the scanner should walk this device now
     */
    public static function consider(array $payload, int $targets_id, int $inventoryInterval): bool
    {
        /** @var DBmysql $DB */
        global $DB;

        $deviceId = trim((string) ($payload['deviceid'] ?? ''));
        if ($deviceId === '') {
            return false;
        }

        $markers = (array) ($payload['markers'] ?? []);
        $now     = time();

        $row = null;
        foreach ($DB->request([
            'FROM'  => self::TABLE,
            'WHERE' => ['deviceid' => $deviceId],
            'LIMIT' => 1,
        ]) as $found) {
            $row = $found;
        }

        $wanted = self::shouldInventory($row, $markers, $now, $inventoryInterval)
            || !self::assetStillExists($row);

        $DB->updateOrInsert(self::TABLE, [
            'plugin_glpinetscan_targets_id' => $targets_id,
            'address'                       => substr((string) ($payload['scan_address'] ?? ''), 0, 63),
            'last_discovery'                => $now,
            'uptime_ticks'                  => (int) ($markers['uptime_ticks'] ?? 0),
            'iftable_last_change'           => (int) ($markers['iftable_last_change'] ?? 0),
            'ifstack_last_change'           => (int) ($markers['ifstack_last_change'] ?? 0),
            'entity_last_change'            => (int) ($markers['entity_last_change'] ?? 0),
        ], ['deviceid' => $deviceId]);

        return $wanted;
    }

    /**
     * The decision itself, given what was recorded last time.
     *
     * Public and free of any database access so it can be tested directly —
     * these four rules are the whole feature, and getting one wrong means
     * either a scanner that never refreshes or one that never saves anything.
     *
     * @param array<string,mixed>|null $row     the previous discovery, if any
     * @param array<string,mixed>      $markers the counters just reported
     */
    public static function shouldInventory(?array $row, array $markers, int $now, int $inventoryInterval): bool
    {
        // Never seen, or seen but never walked.
        if ($row === null || (int) $row['last_inventory'] <= 0) {
            return true;
        }

        // Bounded staleness. Zero or less disables the interval rule entirely,
        // leaving only the change counters — which is a deliberate choice an
        // operator can make for a range of appliances that never change, and a
        // bad one for a switch estate.
        if ($inventoryInterval > 0 && ($now - (int) $row['last_inventory']) >= $inventoryInterval) {
            return true;
        }

        $uptime = (int) ($markers['uptime_ticks'] ?? 0);
        $previous = (int) $row['uptime_ticks'];

        // A reboot. Uptime going backwards is the signal, and it matters twice
        // over: the device may have come back different, and every other
        // counter here is measured against sysUpTime, so they are incomparable
        // across the restart and cannot be trusted until after a walk.
        if ($uptime > 0 && $previous > 0 && $uptime < $previous) {
            return true;
        }

        foreach (['iftable_last_change', 'ifstack_last_change', 'entity_last_change'] as $key) {
            $current = (int) ($markers[$key] ?? 0);
            if ($current > 0 && $current !== (int) $row[$key]) {
                return true;
            }
        }

        return false;
    }

    /**
     * Whether the asset the last full walk produced is still in GLPI.
     *
     * Checked against what that walk actually created, recorded at the time,
     * rather than against `glpi_agents`: GLPI creates no agent row for a device
     * inventoried this way, so a deviceid lookup there never matches and every
     * device would be re-walked on every pass — which is exactly what happened
     * before this was written that way.
     *
     * The case it catches is an operator deleting an asset to force a clean
     * re-import. Without it the scanner would keep discovering the device,
     * decide nothing had changed, and never rebuild what was deleted — the
     * failure mode that makes people stop trusting an inventory tool.
     */
    private static function assetStillExists(?array $row): bool
    {
        if ($row === null) {
            return false;
        }

        $itemtype = (string) ($row['itemtype'] ?? '');
        $items_id = (int) ($row['items_id'] ?? 0);

        // Nothing recorded yet — an older row, or a device never walked. The
        // caller's other rules decide; claiming the asset is missing here would
        // force a walk that may not be needed.
        if ($itemtype === '' || $items_id <= 0) {
            return true;
        }

        if (!class_exists($itemtype) || !is_a($itemtype, \CommonDBTM::class, true)) {
            return false;
        }

        $item = new $itemtype();
        if (!$item->getFromDB($items_id)) {
            return false;
        }

        return (int) ($item->fields['is_deleted'] ?? 0) === 0;
    }

    /** Record that a device was walked in full, and what that produced. */
    public static function recordInventory(string $deviceId, ?string $itemtype, ?int $items_id): void
    {
        /** @var DBmysql $DB */
        global $DB;

        if ($deviceId === '') {
            return;
        }

        $fields = ['last_inventory' => time()];
        if ($itemtype !== null && $itemtype !== '' && $items_id !== null && $items_id > 0) {
            $fields['itemtype'] = $itemtype;
            $fields['items_id'] = $items_id;
        }

        $DB->updateOrInsert(self::TABLE, $fields, ['deviceid' => $deviceId]);
    }

    /**
     * Forget everything known about a target's devices.
     *
     * Used when a target's ranges change: the devices behind it may not be the
     * same devices, and a stale "last walked" would suppress the walk that
     * would have found that out.
     */
    public static function forgetTarget(int $targets_id): void
    {
        /** @var DBmysql $DB */
        global $DB;

        $DB->delete(self::TABLE, ['plugin_glpinetscan_targets_id' => $targets_id]);
    }
}
